x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/build_exceptions_query.ts

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { Query } from '../../../../../../../../src/plugins/data/server';
+import { List, ListOperator, ListValues } from '../routes/schemas/types/lists_default_array';
+import { RuleAlertParams, Language } from '../types';
+
+interface Operators {
+  and: string;
+  not: string;
+  or: string;
+}
+
+export const getLanguageBooleanOperator = ({
+  language,
+  value,
+}: {
+  language: keyof Language;
+  value: keyof Operators;
+}) => {
+  if (language === 'lucene') {
+    return value.toUpperCase();
+  } else {
+    return value;
+  }
+};
+
+export const operatorBuilder = ({
+  operator,
+  language,
+}: {
+  operator: ListOperator;
+  language: keyof Language;
+}): string => {
+  return operator === 'excluded'
+    ? ` ${getLanguageBooleanOperator({ language, value: 'and' })} `
+    : ` ${getLanguageBooleanOperator({ language, value: 'and' })} ${getLanguageBooleanOperator({
+        language,
+        value: 'not',
+      })} `;
+};
+
+export const buildExists = ({
+  operator,
+  field,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  language: keyof Language;
+}): string => {
+  switch (language) {
+    case 'kuery':
+      return `${operatorBuilder({ operator, language })}${field}:*`;
+    case 'lucene':
+      return `${operatorBuilder({ operator, language })}_exists_${field}`;
+    default:
+      return '';
+  }
+};
+
+export const buildMatch = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: keyof Language;
+}) => {
+  const value = values[0].name;
+  return `${operatorBuilder({ operator, language })}${field}:${value}`;
+};
+
+export const buildMatchAll = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: keyof Language;
+}) => {
+  if (values.length === 1) {
+    return buildMatch({ operator, field, values, language });
+  } else {
+    const matchAllValues = values.map(value => {
+      return value.name;
+    });
+
+    return `${operatorBuilder({ operator, language })}${field}:(${matchAllValues.join(
+      ` ${getLanguageBooleanOperator({ language, value: 'or' })} `
+    )})`;
+  }
+};
+
+export const evaluateValues = ({
+  list,
+  language,
+}: {
+  list: List;
+  language: keyof Language;
+}): string => {
+  const { values_operator: operator, values_type: type, field, values } = list;
+  switch (type) {
+    case 'exists':
+      return buildExists({ operator, field, language });
+    case 'match':
+      return buildMatch({ operator, field, values: values!, language });
```

Comment:
You really should avoid `void non-null assertions` as per the style guidelines:
https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md#avoid-non-null-assertions

I am going to turn that linter rule on very soon for detection engine and other areas. It has been the source of so many bugs and confusing readability.

Since we have better operators such as `??` you can do:

```ts
    case 'match':
      return buildMatch({ operator, field, values: values ?? [], language });
    case 'match_all':
      return buildMatchAll({ operator, field, values: values ?? [], language });
```

And that will pass the linter and keep things from unexpectedly blowing up later. You can do that or you can flow downstream to `buildMatchAll` and push the undefined and null and manage it there. Either way, but turning off type assertions is best avoided at all costs.

Reply:
Went through and tried to address any areas where it was not following the style guidelines. Removed the use of `!`. Thanks for catching that.